File: src/lib/preview.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

/**
 * Mirrors `crate::preview::DetectedServer` on the Rust side.
 * `title` is the parsed contents of the page's `<title>` tag when present.
 */
export interface DetectedServer {
  port: number;
  url: string;
  title: string | null;
}

/**
 * Run a one-shot port sweep right now. Used when the preview tab opens so
 * the UI doesn't have to wait up to 3s for the watcher's next tick.
 */
export async function listDevServers(): Promise<DetectedServer[]> {
  return invoke<DetectedServer[]>("list_dev_servers");
}

/**
 * Start the background dev-server poll loop. Safe to call repeatedly —
 * any existing watcher is replaced. The watcher itself is also kicked off
 * automatically from the Tauri `setup` hook, so most callers won't need
 * this.
 */
export async function startPreviewWatcher(): Promise<void> {
  await invoke<void>("start_preview_watcher");
}

/** Stop the background poll loop. Resolves to `true` if one was running. */
export async function stopPreviewWatcher(): Promise<boolean> {
  return invoke<boolean>("stop_preview_watcher");
}

/**
 * Subscribe to `preview:servers` events. The payload is the full live list
 * of detected dev servers — replace any local state on each event.
 * Returns an unlisten function.
 */
export async function subscribePreviewServers(
  cb: (servers: DetectedServer[]) => void,
): Promise<UnlistenFn> {
  return listen<DetectedServer[]>("preview:servers", (evt) => cb(evt.payload));
}

File: src-tauri/src/commands/preview.rs
//! Tauri commands for the localhost dev-server preview pane.
//!
//! Thin shim around `crate::preview` — the heavy lifting (port sniffing,
//! `<title>` parsing, event emission) lives in that module. These commands
//! exist so the React frontend can drive the watcher and trigger one-shot
//! sweeps.

use crate::preview::{self, DetectedServer};

/// Run a single synchronous sweep of the probe ports. Useful when the
/// preview tab first opens — the watcher emits *changes* only, so without
/// this the UI would have to wait up to 3s for the first event.
#[tauri::command]
pub async fn list_dev_servers() -> Result<Vec<DetectedServer>, String> {
    Ok(preview::sweep_once().await)
}

/// Start the background poll loop. Safe to call repeatedly — any existing
/// watcher is replaced.
#[tauri::command]
pub async fn start_preview_watcher(app: tauri::AppHandle) -> Result<(), String> {
    preview::start(app).map_err(|e| format!("start_preview_watcher: {e:#}"))
}

/// Stop the background poll loop if running. Returns `true` if a watcher
/// was actually stopped.
#[tauri::command]
pub async fn stop_preview_watcher() -> Result<bool, String> {
    Ok(preview::stop())
}
